walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,128 @@
+
+        // Remove the last added timerange to consider the next event. 
+        availableTimes.remove(availableTimes.size() - 1);
+
+        // If one event fully contains another (nested event), there are 2 possible times: before and after the longest event. 
+        if (previousEvent.getWhen().end() > currentEvent.getWhen().start()
+            && previousEvent.getWhen().end() > currentEvent.getWhen().end()) {
+          availableTimes.add(TimeRange.fromStartEnd(previousEvent.getWhen().end(), TimeRange.END_OF_DAY, true));
+          break;
+        }
+
+        // If events have no conflict, there are 3 possible times: before the first one, between the 2 events and after the second one.
+        long durationBetweenEvents = currentEvent.getWhen().start() - previousEvent.getWhen().end(); 
+        if (request.getDuration() <= durationBetweenEvents) 
+          availableTimes.add(TimeRange.fromStartEnd(previousEvent.getWhen().end(), currentEvent.getWhen().start(), false));
+
+        // If the end of the event is not the end of the day, then the time in between is a possible time. 
+        if (currentEvent.getWhen().end() != TimeRange.END_OF_DAY + 1) {
+          availableTimes.add(TimeRange.fromStartEnd(currentEvent.getWhen().end(), TimeRange.END_OF_DAY, true));
+        }
+      }
+
+      // If there were no previous times, then this event is the first one of the day. The possible times are before and after the event.
+      else {
+        // If the first event of the day starts at the beginning of the day, there are no availabilities before it. 
+        if (currentEvent.getWhen().start() != TimeRange.START_OF_DAY) 
+          availableTimes.add(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, currentEvent.getWhen().start(), false));
+  
+        if (currentEvent.getWhen().end() != TimeRange.END_OF_DAY + 1)
+          availableTimes.add(TimeRange.fromStartEnd(currentEvent.getWhen().end(), TimeRange.END_OF_DAY, true));
+      }
+    }
+    return availableTimes;
+  }
+
+  //Return the events common to both required and optional attendees
+  public Collection<TimeRange> getCommonTimes(List<Event> eventsWithRequiredAttendees,
+      List<Event> eventsWithOptionalAttendees, MeetingRequest request) {
+    
+    Collection<TimeRange> requiredAttendeesAvailabilities = getPossibleTimes(eventsWithRequiredAttendees, request);
+    Collection<TimeRange> optionalAttendeesAvailabilities = getPossibleTimes(eventsWithOptionalAttendees, request);
+    Collection<TimeRange> commonAvailabilities = new ArrayList<>();
+
+    if (optionalAttendeesAvailabilities.isEmpty())
+      return requiredAttendeesAvailabilities;
+
+    for (TimeRange requiredAttendeesTimerange : requiredAttendeesAvailabilities) {
+      for (TimeRange optionalAttendeesTimerange : optionalAttendeesAvailabilities) {
+        if (requiredAttendeesTimerange.overlaps(optionalAttendeesTimerange))
+          commonAvailabilities.add(requiredAttendeesTimerange);
```

Comment:
I don't know enough about the inputs here - is this trying to take the intersection of the required and optional time ranges? If so, does it work if you have inputs like:
```
   |------|
       |-------|
```
?

If this isn't quite working as intended, you could e.g. create a TimeRange using "fromStartEnd" with the min start and the max end.

Reply:
Thanks @DouglasTurk! Your feedback was very helpful. I realised that overlaps() might not be ideal to find the common times. I ended up using the contains() function (which, similarly to overlaps(), was already written for us in the TimeRange java file). My code should now check the intersection between a required and optional event. I also added more tests to see whether it passes the scenarios you have proposed here.